### DISCUSSION TITLE
BiC Problem

### DISCUSSION THREAD

**Discussion by Sean3216** [ContinualAI/avalanche#1404] [Q&A]
Hi. I would like to ask about BiC implementation using Avalanche. I tried to run it but I encounter the following error:

------------------------------------------------------------------------------------------------------------------------------------------------------
![image](https://github.com/ContinualAI/avalanche/assets/92766104/8bb6ec99-ce18-47b4-bf5f-f2ba150764e3)
------------------------------------------------------------------------------------------------------------------------------------------------------

This is how I implement the BiC using in the main.py and learning.py:

------------------------------------------------------------------------------------------------------------------------------------------------------
main.py:
```
for task in tasknumbers:
        train.append(alltaskdataloader[task]['train'])
        test.append(alltaskdataloader[task]['test'])
    
    scenario_stream = Generator.dataset_benchmark(train, test)
    model = NetCifar()
    CL = Continual_Learning(model)
    CL.bic(scenario_stream)
```

learning.py:
```
def bic(self, stream):
        self.checkpoint_plugin = CheckpointPlugin(
            FileSystemCheckpointStorage(
                directory = './checkpoints/bic'
            ),
            map_location = self.device
        )
        cl_strategy, initial_exp = self.checkpoint_plugin.load_checkpoint_if_exists()
        if cl_strategy is None:
            plugins = [self.checkpoint_plugin]

            os.makedirs(f'./logs/checkpointing/', exist_ok = True)
            loggers = [
                InteractiveLogger(),
                TensorboardLogger(f'./logs/checkpointing')
            ]
            eval_plugin = EvaluationPlugin(accuracy_metrics(minibatch=False, epoch=True, experience=True, stream=True),
                                            loss_metrics(minibatch=False, epoch=True, experience=True, stream=True),
                                            timing_metrics(epoch=True, epoch_running=True),
                                            forgetting_metrics(experience=True, stream=True),
                                            loggers=loggers
                                            )
            cl_strategy = BiC(self.modelcl, torch.optim.SGD(self.modelcl.parameters(), lr = 0.001, momentum = 0.9),
                                nn.CrossEntropyLoss(), mem_size = 500, val_percentage = 0.2, T = 5, stage_2_epochs = 5, 
                                lamb = 0.5, lr = 0.01, device = self.device, train_mb_size = 64,
                                train_epochs = 5, eval_mb_size = 64, plugins = plugins, evaluator = eval_plugin)
        print("Starting experiment...")
        results = []
        for experience in stream.train_stream[initial_exp:]:
            print("Start of Training for Task ", experience.current_experience)
            cl_strategy.train(experience)
            print("Training completed")
            print("Computing accuracy on the whole test set")
            results.append(cl_strategy.eval(stream.test_stream)) 
```

------------------------------------------------------------------------------------------------------------------------------------------------------
            

Why did an error like that pop out? Thank you for the time to read this long question.

**Comment by JuliousHurtado** [comment]
Hi @Sean3216

Let me see if I can help. 
Unfortunately with the current information I can't replicate your error.

I few questions to help me identify the problem:
- Are you using the Avalanche implementation of BiC?
- Have you try run BiC without the checkpoint plugin?
- Have you try using the Naive strategy and add BiC as a plugin (BiCPlugin)?
- Which version of avalanche are you using?

Thanks!
